S2A record protocol implementation, part 2

matthewstevenson88 opened the pull request:
This PR contains the implementation of the "encrypt" functions for the S2A record protocol, along with the associated unit tests. Note that the only files that need to be reviewed are the following:
- alts_iovec_record_protocol.{h,cc}
- s2a_crypter.{h,cc}
- s2a_test_util.{h,cc}
- s2a_record_protocol_test.cc

A few notes:
1. The `s2a_write_tls13_record` method is exposed because it is needed in the S2A version of the zero-copy frame protector, whose implementation will follow in a separate PR. 
2. This PR is being merged into the `s2a-rp-create` branch (rather than `s2a-master`) because of small build fixes that are added in `s2a-rp-create`; once #4 is merged, the target will change to `s2a-master`.

cesarghali commented on test/core/tsi/s2a/s2a_test_util.cc:
```diff
@@ -185,3 +187,220 @@ size_t expected_message_size(size_t plaintext_size) {
+        "If |record_three| is nullptr, then |record_three_size| must be zero.");
+    return false;
+  }
+  switch (ciphersuite) {
```
These three `switch` statements seem to have very similar logic. Can we have that logic in a separate function that we can three times here?

matthewstevenson88 replied:
I turned the body of the cases into a standalone function, `verify_record`, that is called in each case with different parameters.

cesarghali commented on test/core/tsi/s2a/s2a_test_util.cc:
```diff
@@ -185,3 +187,220 @@ size_t expected_message_size(size_t plaintext_size) {
+       *   {23, 3, 3, 0, 23, 242, 228, 228, 17, 172, 103, 96, 228, 227,
+       *   240, 116, 163, 101, 116, 196, 94, 228, 193, 144, 97, 3, 219, 13}.
+       */
+      uint8_t correct_record_one_bytes[28] = {
```
Please define hard coded buffers like these as const at the top of the file.

cesarghali replied:
Also, out of curiosity, how is this generated?

matthewstevenson88 replied:
Done. These examples were generated using sslcryptor.

cesarghali commented on test/core/tsi/s2a/s2a_test_util.cc:
```diff
@@ -185,3 +187,220 @@ size_t expected_message_size(size_t plaintext_size) {
+        gpr_strdup("If |record| is nullptr, then |record_size| must be zero.");
+    return false;
+  }
+  switch (ciphersuite) {
```
We can use the function I mentioned above to test this logic as well.

matthewstevenson88 replied:
Done.

cesarghali commented on test/core/tsi/s2a/s2a_test_util.h:
```diff
@@ -38,4 +38,12 @@ grpc_byte_buffer* create_example_session_state(bool admissible_tls_version,
 
 size_t expected_message_size(size_t plaintext_size);
 
+bool check_encrypt_record(TLSCiphersuite ciphersuite, uint8_t* record_one,
```
Can you please add a comment explaining what this function and the one below do? Thanks!

matthewstevenson88 replied:
Done.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +257,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+ * function returns zero.
+ *  - crypter: an instance of s2a_crypter. **/
+static size_t s2a_tag_size(const s2a_crypter* crypter) {
+  if (crypter == nullptr) {
```
Done in next commit.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +257,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+ *  TLS 1.3 record that is handled by |crypter|. If |crypter| is nullptr, then
+ *  this function returns zero.
+ *  - crypter: an instance of s2a_crypter. **/
+static size_t s2a_max_aead_nonce_size(const s2a_crypter* crypter) {
```
Changed to `GPR_ASSERT(crypter != nullptr)`. 

Also, you are correct that this function may become more complex if additional ciphersuites are added. I've added a comment to clarify that.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +257,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+ *  |half_connection| is nullptr, then the function returns
+ *  GRPC_STATUS_INTERNAL. Otherwise, the function returns GRPC_STATUS_OK.
+ *  - half_connection: an instance of s2a_half_connection. **/
+grpc_status_code increment_sequence(s2a_half_connection* half_connection) {
```
Done.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +257,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+ *  - out_bytes: a pointer to a length 8 array of bytes. **/
+static void sequence_to_bytes(const s2a_half_connection* half_connection,
+                              uint8_t* out_bytes) {
+  if (half_connection != nullptr && out_bytes != nullptr) {
```
Done.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +257,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+}
+
+size_t s2a_max_record_overhead(const s2a_crypter* crypter) {
+  if (crypter == nullptr || crypter->out_connection == nullptr ||
```
Done.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +257,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+      !crypter->out_connection->initialized) {
+    return 0;
+  }
+  return SSL3_RT_HEADER_LENGTH + s2a_tag_size(crypter) + 1;
```
Done.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +257,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+}
+
+static grpc_status_code s2a_write_tls13_record_header(uint8_t record_type,
+                                                      uint8_t* record_header,
```
Done.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +257,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+                                                      size_t header_size,
+                                                      size_t payload_size,
+                                                      char** error_details) {
+  if (record_header == nullptr) {
```
Done.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +257,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+    s2a_crypter* crypter, uint8_t record_type, const iovec* unprotected_vec,
+    size_t unprotected_vec_size, iovec protected_record, size_t* bytes_written,
+    char** error_details) {
+  if (crypter == nullptr || crypter->out_connection == nullptr ||
```
Done.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +257,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+  if (encrypt_status != GRPC_STATUS_OK) {
+    return encrypt_status;
+  }
+  GPR_ASSERT(bytes_written != nullptr);
```
Done.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +257,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+        "record.");
+    return GRPC_STATUS_FAILED_PRECONDITION;
+  }
+  if (protected_record.iov_base == nullptr || protected_record.iov_len == 0) {
```
Added an assertion that `protected_record.iov_base != nullptr` at the beginning. The case where `protected_record.iov_len = 0` is actually handled by the previous if statement.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +257,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+  /** Set up the additional_data_bytes buffer, which is to be authenticated
+   *  but not encrypted. **/
+  uint8_t additional_data_bytes[sizeof(sequence) + SSL3_RT_HEADER_LENGTH];
+  GPR_ASSERT(sizeof(additional_data_bytes) == 13);
```
Done.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +257,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+  size_t nonce_size = crypter->out_connection->fixed_nonce_size;
+  memset(nonce_storage, 0, sizeof(nonce_storage));
+  nonce = nonce_storage;
+  memcpy(nonce, crypter->out_connection->fixed_nonce, nonce_size);
```
Done.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +257,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+        gpr_strdup("The input |crypter| was not properly initialized.");
+    return GRPC_STATUS_FAILED_PRECONDITION;
+  }
+  if ((unprotected_vec == nullptr && unprotected_vec_size != 0) ||
```
Changed to an if statement wrapping an assert.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +257,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+      return EVP_AEAD_AES_GCM_TAG_LEN;
+    case TLS_CHACHA20_POLY1305_SHA256:
+      return POLY1305_TAG_LEN;
+    default:
```
Sure thing. The default case now logs an error and aborts.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +257,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+  uint8_t sequence[8];
+  sequence_to_bytes(crypter->out_connection, sequence);
+
+  /** Set up the additional_data_bytes buffer, which is to be authenticated
```
Done.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +257,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+  additional_data += sizeof(sequence);
+  iovec aad_vec = {(void*)additional_data, SSL3_RT_HEADER_LENGTH};
+
+  /** The following constructs the nonce for the TLS payload in local
```
Done.

matthewstevenson88 commented on test/core/tsi/s2a/s2a_record_protocol_test.cc:
```diff
@@ -245,12 +246,268 @@ static void test_create_crypter_success(TLSCiphersuite ciphersuite) {
+    grpc_core::Delete<grpc_channel>(channel);
+    return;
+  }
+  GPR_ASSERT(status == GRPC_STATUS_OK);
```
Unfortunately I can't move these 2 asserts before the if statement. The reason is that, if the CHACHA-POLY ciphersuite is used, then e.g. `status` will not be `GRPC_STATUS_OK`, rather it will be `GRPC_STATUS_UNIMPLEMENTED`. I don't want the test to fail in the case that the CHACHA-POLY ciphersuite is used.

matthewstevenson88 commented on test/core/tsi/s2a/s2a_record_protocol_test.cc:
```diff
@@ -245,12 +246,268 @@ static void test_create_crypter_success(TLSCiphersuite ciphersuite) {
+  GPR_ASSERT(error_details == nullptr);
+
+  /** Test the case when the memory allocated for the record is insufficient.**/
+  size_t test_plaintext_size = 7;
```
Done here and in subsequent tests.

matthewstevenson88 commented on test/core/tsi/s2a/s2a_record_protocol_test.cc:
```diff
@@ -245,12 +246,268 @@ static void test_create_crypter_success(TLSCiphersuite ciphersuite) {
+  /** Test the case when the memory allocated for the record is insufficient.**/
+  size_t test_plaintext_size = 7;
+  uint8_t test_plaintext[7] = "123456";
+  uint8_t* test_plaintext_pointer = test_plaintext;
```
Done.

matthewstevenson88 commented on test/core/tsi/s2a/s2a_record_protocol_test.cc:
```diff
@@ -245,12 +246,268 @@ static void test_create_crypter_success(TLSCiphersuite ciphersuite) {
+                  record_allocated_size, &record_size, &error_details);
+  GPR_ASSERT(insufficient_memory_status == GRPC_STATUS_FAILED_PRECONDITION);
+  int correct_error_message = strcmp(error_details,
+                                     "The plaintext size is too large to fit "
```
Done; added the constants to `s2a_constants.h`.

matthewstevenson88 commented on test/core/tsi/s2a/s2a_test_util.cc:
```diff
@@ -53,7 +55,7 @@ grpc_byte_buffer* create_example_session_state(bool admissible_tls_version,
       0xda, 0xc7, 0x31, 0xae, 0x48, 0x66, 0x67, 0x7e, 0xd2, 0xf6, 0x5c,
       0x49, 0x0e, 0x18, 0x81, 0x7b, 0xe5, 0xcb, 0xbb, 0xd0, 0x3f, 0x59,
-      0x7a, 0xd5, 0x90, 0x41, 0xc1, 0x17, 0xb7, 0x32, 0x10, 0x9a, 0x00};
+      0x7a, 0xd5, 0x90, 0x41, 0xc1, 0x17, 0xb7, 0x31, 0x10, 0x9a, 0x00};
```
This was a typo from when I copied the data over. (I took the key used by sslcryptor and plugged it into a decimal->hex converter, and then hardcoded them here. Also, from the point of view of changing test vectors, this is the key, so we should be able to enter whatever we like here.)

matthewstevenson88 commented on test/core/tsi/s2a/s2a_test_util.cc:
```diff
@@ -185,3 +187,220 @@ size_t expected_message_size(size_t plaintext_size) {
+                          size_t record_one_size, uint8_t* record_two,
+                          size_t record_two_size, uint8_t* record_three,
+                          size_t record_three_size, char** error_details) {
+  if (record_one == nullptr && record_one_size != 0) {
```
Done.

cesarghali commented on test/core/tsi/s2a/s2a_test_util.cc:
```diff
@@ -17,10 +17,84 @@
 #include "src/core/tsi/s2a/record_protocol/s2a_crypter_util.h"
 #include "src/core/tsi/s2a/s2a_constants.h"
 
+uint8_t aes_128_gcm_key_bytes[17] = {0xc3, 0xae, 0x75, 0x09, 0xcf, 0xce,
```
Can you please add a comment about how these vectors were generated in case we need to regenerate them in the future.

matthewstevenson88 replied:
Done.

matthewstevenson88 commented on test/core/tsi/s2a/s2a_test_util.h:
```diff
@@ -38,4 +38,31 @@ grpc_byte_buffer* create_example_session_state(bool admissible_tls_version,
+
+/** This method returns true if |record| and |record_size| match the TLS 1.3
+ *  record obtained from an empty plaintext and using a crypter configured by
+ * the output of the |create_example_session_state| method. Otherwise, this
```
Done in next commit.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +250,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+  return additional_data;
+}
+
+static uint8_t* s2a_nonce(s2a_crypter* crypter, uint8_t* sequence,
```
Done.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -357,48 +344,67 @@ static grpc_status_code s2a_write_tls13_record_header(uint8_t record_type,
   return GRPC_STATUS_OK;
 }
 
+static uint8_t* s2a_additional_data(uint8_t* sequence, size_t sequence_size,
```
Done.

matthewstevenson88 commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +250,226 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+  GPR_ASSERT(crypter->out_connection->initialized);
+  GPR_ASSERT(protected_record.iov_base != nullptr);
+  GPR_ASSERT(bytes_written != nullptr);
+  if (unprotected_vec == nullptr || unprotected_vec_size == 0) {
```
That's right: `unprotected_vec` is `nullptr` iff `unprotected_vec_size = 0`.

I've changed it back to an if statement closer to what it was before.

cesarghali replied:
Do we still need this if and assert?

matthewstevenson88 replied:
Good catch. I removed the assert.

matthewstevenson88 commented on test/core/tsi/s2a/s2a_record_protocol_test.cc:
```diff
@@ -245,12 +244,259 @@ static void test_create_crypter_success(TLSCiphersuite ciphersuite) {
+  GPR_ASSERT(record_one_size ==
+             expected_message_size(test_plaintext_one_size - 1));
+
+  const size_t test_plaintext_two_size = 10;
```
I've added a utility function that captures the encryption and the subsequent `GPR_ASSERT`s. I didn't wrap the initialization of the plaintext and record into the utility function, however, because all the records are needed for verification after all encryption is done. 

If you had something else in mind, let me know!

cesarghali replied:
SG, thanks!

matthewstevenson88 commented:
Thanks @cesarghali for the review! Fixed the nits in the most recent push.

@beschmi Can you have a look at the PR? (And also #4.) Thanks!

beschmi commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -66,21 +67,19 @@ static grpc_status_code assign_crypter(bool in, uint8_t* derived_key,
+    case kTlsAes256GcmSha384:
       aead_crypter_status = gsec_aes_gcm_aead_crypter_create(
           derived_key, key_size, nonce_size, tag_size,
           /** rekey **/ false, &aead_crypter, error_details);
```
Nit: /* rekey= */

matthewstevenson88 replied:
Done.

beschmi commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -150,38 +148,32 @@ grpc_status_code s2a_crypter_create(
   }
 
-  size_t tag_size = (rp_crypter->ciphersuite == TLS_CHACHA20_POLY1305_SHA256)
+  size_t tag_size = (rp_crypter->ciphersuite == kTlsChacha20Poly1305Sha256)
```
Can you call s2a_tag_size here?

matthewstevenson88 replied:
I can - good catch! Switched this.

beschmi commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +248,244 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+  size_t additional_data_size = sequence_size + header_size;
+  GPR_ASSERT(additional_data_size == kTlsAdditionalDataBytesSize);
+  uint8_t* additional_data =
+      (uint8_t*)gpr_malloc(additional_data_size * sizeof(uint8_t));
```
Can you use reinterpret_cast<uint*>(...)?

beschmi replied:
Or static_cast ...

matthewstevenson88 replied:
Changed (and the other similar instances of this).

beschmi commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +248,244 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+  GPR_ASSERT(protected_record.iov_base != nullptr);
+  GPR_ASSERT(bytes_written != nullptr);
+  if ((unprotected_vec == nullptr && unprotected_vec_size > 0) ||
+      (unprotected_vec != nullptr && unprotected_vec_size == 0)) {
```
Not sure why you need this second check, having non-null pointer and size == 0 is pretty natural.

matthewstevenson88 replied:
Removed the second check.

beschmi commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +248,244 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+  iovec aad_vec = {(void*)(additional_data + 8), SSL3_RT_HEADER_LENGTH};
+
+  /** The following constructs the nonce for the TLS payload in local
+   *  storage. It is built by taking the fixed nonce from |crypter|'s
```
s/fixed nonce/nonce mask/?

matthewstevenson88 replied:
Done.

matthewstevenson88 commented:
Thanks for the review @beschmi! Implemented the changes you requested, and also added the std::vector to the unit tests (in the same way as in #8).

@yihuazhang Could you take a look now to make sure it follows the gRPC core style?

yihuazhang commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -55,6 +56,23 @@ typedef struct s2a_crypter {
+      return kPoly1305TagLength;
+    default:
+      gpr_log(GPR_ERROR, kS2AUnsupportedCiphersuite);
+      abort();
```
`abort()` seems a little strong to me. Can we handle it more gracefully?

matthewstevenson88 replied:
In the most recent commit (to come), I refactored this method to output a status code and populate some error details in this case.

yihuazhang commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +266,227 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+  size_t additional_data_size = sequence_size + header_size;
+  GPR_ASSERT(additional_data_size == kTlsAdditionalDataBytesSize);
+  uint8_t* additional_data =
+      static_cast<uint8_t*>(gpr_malloc(additional_data_size * sizeof(uint8_t)));
```
s/gpr_malloc/gpr_zalloc

matthewstevenson88 replied:
Done.

yihuazhang commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +266,227 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+  GPR_ASSERT(nonce_size != nullptr);
+  uint8_t* nonce =
+      static_cast<uint8_t*>(gpr_malloc(max_nonce_size * sizeof(uint8_t)));
+  *nonce_size = crypter->out_connection->fixed_nonce_size;
```
How about adding GPR_ASSERT for `crypter->out_connection != nullptr`?

matthewstevenson88 replied:
Good call. Done.

yihuazhang commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -66,21 +84,19 @@ static grpc_status_code assign_crypter(bool in, uint8_t* derived_key,
-      *error_details = gpr_strdup(
-          "The crypter's ciphersuite is not supported; cannot initialize AEAD "
-          "crypter.");
+      *error_details = gpr_strdup(kS2AUnsupportedCiphersuite);
```
The memory has not been freed by the caller.

matthewstevenson88 replied:
Per our discussion, I left this is as (so that the caller of these methods has a consistent expectation for how to free `error_details`).

yihuazhang commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +266,227 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+  GPR_ASSERT(crypter != nullptr);
+  /** If additional supported ciphersuites are added, then there may be
+   *  additional options for the return value. **/
+  return kEvpAeadMaxNonceLength;
```
It seems a little weird in a sense that `kEvpAeadMaxNonceLength` is not a part of `s2a_crypter`. Do we need this function?

matthewstevenson88 replied:
I agree it looks a little strange right now, but it's because we might add other ciphersuites or other TLS versions, in which case the max nonce length would change. I wrote it this way so the API would be stable.

yihuazhang commented on src/core/tsi/s2a/record_protocol/s2a_crypter.cc:
```diff
@@ -256,3 +266,227 @@ void check_half_connection(s2a_crypter* crypter, bool in_half_connection,
+
+  half_connection->sequence += 1;
+  /** Check whether the sequence number has overflowed. **/
+  if (half_connection->sequence == 0) {
```
I think it's fine to use the current check since uint64_t will always occupy 8 bytes across different platforms.

matthewstevenson88 replied:
Sounds good, left it as is.

yihuazhang commented:
Looks good, Matt. I left some minor comments.

matthewstevenson88 commented:
Thanks for the quick review @yihuazhang! All your comments were fixed in the most recent commit.

matthewstevenson88 commented:
Thanks @yihuazhang for the review! :)
